File: emulator.zig
const std = @import("std");
const expect = std.testing.expect;
const print = std.debug.print;

const debug = false;
const max_run_len = 100;

const memsize = 1 << 8;
const Word = u8;
const SWord = i8;

// Instruction layout:
// IIiii SAA
// I = Instruction kind
// i = Instruction
// S = self-modify / zero flag
// A = dereference args

const Instruction = packed struct {
    deref1: bool,
    deref0: bool,
    self_modify: bool,
    operation: Operation,

    pub fn argc(self: Instruction) u2 {
        const num_args: u2 = switch (self.operation.kind()) {
            .one_arg => 1,
            .two_arg, .unary => 2,
            .binary => 3,
        };
        return num_args - @intFromBool(self.self_modify);
    }

    pub fn print_info(self: Instruction) void {
        print("op:{s} d0:{d} d1:{d} sm:{d}", .{
            @tagName(self.operation),
            @intFromBool(self.deref0),
            @intFromBool(self.deref1),
            @intFromBool(self.self_modify)});
    }

    pub fn has_writeback(self: Instruction) bool {
        return self.operation.kind() == Operation.Kind.binary or
            self.operation.kind() == Operation.Kind.unary;
    }

    pub fn has_two_inputs(self: Instruction) bool {
        return self.operation.kind() == Operation.Kind.two_args or
            (self.operation.kind() == Operation.Kind.binary and !self.self_modify);
    }
};

test "Instruction is 8-bit" {
    try expect(@sizeOf(Instruction) == 1);
}

const Operation = enum(u5) {
    // -- Bank 0: X = Y
    peek = 0b00_000,
    copy,
    deref,
    getflag,
    pop,
    not,
    negate,
    invert,

    // -- Bank 1: OP X Y
    poke = 0b01_000,
    jumpifz,
    jumpifnz,
    setflag,
    // -- Bank 1.2: OP X
    push,
    call,
    ret,
    // unused,

    // -- Bank 2: X = Y OP Z
    add = 0b10_000,
    sub,
    // unused,
    // unused,
    b_and = 0b10_100,
    b_or,
    b_xor,
    // unused,

    // -- Bank 3: X = Y OP Z (2)
    u_shiftl = 0b11_000,
    u_shiftr,
    s_shiftr,
    // unused,
    u_gt = 0b11_100,
    u_lt,
    s_gt,
    s_lt,

    const Kind = enum {
        unary,
        binary,
        one_arg,
        two_arg,
    };

    pub fn kind(self: Operation) Operation.Kind {
        const upper_bank = @intFromEnum(self) & (1 << 2) != 0;
        const kind_bits: u2 = @truncate(@intFromEnum(self) >> 3);
        return switch (kind_bits) {
            0 => Kind.unary,
            1 => if (upper_bank) Kind.one_arg else Kind.two_arg,
            2,3 => Kind.binary,
        };
    }
};

const Cpu = struct {
    pc: Word,
    memory: [memsize]Word,
    halted: bool,

    pub fn init() Cpu {
        return std.mem.zeroInit(Cpu, .{});
    }

    pub fn fetch_next(self: *Cpu) Word {
        const out = self.memory[self.pc];
        self.pc +%= 1;
        return out;
    }

    fn signed(w: Word) SWord {
        return @bitCast(w);
    }

    fn unsigned(w: SWord) Word {
        return @bitCast(w);
    }

    pub fn step(self: *Cpu) void {
        // TODO: implement self modifying bit
        var args: [3]Word = undefined;
        var retval: Word = 0;

        // Fetch
        const ix: Instruction = @bitCast(self.fetch_next());
        if (debug) {
            ix.print_info();
            print("; ", .{});
        }

        for (0..ix.argc()) |i| {
            args[i] = self.fetch_next();
            if (debug) print("param {x:0>2}; ", .{args[i]});
        }
        if (debug) print("\n", .{});

        // Read memory
        for (0..2) |i| {
            const memflag = if (i == 0) ix.deref0 else ix.deref1;
            if (memflag) {
                var idx = switch(ix.operation.kind()) {
                    .unary, .two_arg => i,
                    .binary => i+1,
                    .one_arg => 0,
                };
                if (ix.operation == Operation.deref) idx = 1;
                const addr = args[idx];
                args[idx] = self.memory[addr];
                if (debug) print("read {x:0>2} from {x:0>2}\n", .{args[idx], addr});
            }
        }

        if (ix.self_modify) {
            if (ix.operation.kind() == Operation.)
        }

        // Perform operation
        switch (ix.operation) {
            // Bank 0, unary
            //.peek => {}, // TODO: I/O
            .copy, .deref => retval = args[1],
            //.getflag => {}, // TODO: flags
            //.pop => {}, // TODO: stack
            .not => retval = if (args[1] == 0) 1 else 0,
            .negate => retval = unsigned(-signed(args[1])),
            .invert => retval = ~args[1],

            // Bank 1, two-arg
            .poke => print("{c}", .{args[1]}),
            .jumpifz => { if (args[1] == 0) self.pc = args[0]; },
            .jumpifnz => { if (args[1] != 0) self.pc = args[0]; },
            .setflag => self.halted = true, // TODO flags

            // Bank 1.2, one-arg
            //.push => {}, // TODO stack
            //.call => {}, // TODO stack
            //.ret => {}, // TODO stack

            // Bank 2/3, binary
            .add => retval = args[1] +% args[2], // TODO carry
            .sub => retval = args[1] -% args[2], // TODO carry
            .b_and => retval = args[1] & args[2],
            .b_or => retval = args[1] | args[2],
            .b_xor => retval = args[1] ^ args[2],

            .u_shiftl => retval = args[1] << @truncate(args[2]), // TODO carry
            .u_shiftr => retval = args[1] >> @truncate(args[2]), // TODO carry
            .s_shiftr => retval = unsigned(signed(args[1]) >> @truncate(args[2])), // TODO carry

            .u_gt => retval = if (args[1] > args[2]) 1 else 0,
            .u_lt => retval = if (args[1] < args[2]) 1 else 0,
            .s_gt => retval = if (signed(args[1]) > signed(args[2])) 1 else 0,
            .s_lt => retval = if (signed(args[1]) < signed(args[2])) 1 else 0,
            else => {},
        }

        // Writeback
        if (ix.has_writeback()) {
            self.memory[args[0]] = retval;
            if (debug) print("wrote {x:0>2} to {x:0>2}\n", .{retval, args[0]});
        }
    }

    pub fn run(self: *Cpu) void {
        for (0..max_run_len) |_| {
            self.step();
            if (self.halted) break;
        }
    }
};

pub fn main() !void {
    var cpu = Cpu.init();
    var assembler = Assembler.new(&cpu.memory);
    assembler.write_hello();
    cpu.run();
}

const Assembler = struct {
    addr: Word,
    binary: *[memsize]Word,

    pub fn new(binary: *[memsize]Word) Assembler {
        return Assembler {
            .addr = 0,
            .binary = binary,
        };
    }

    fn push_word(self: *Assembler, word: Word) void {
        self.binary[self.addr] = word;
        self.addr += 1;
    }

    fn push_instruction(self: *Assembler, op: Operation, flags: u3) void {
        const inst = Instruction {
            .operation = op,
            .deref0 = flags & 4 != 0,
            .deref1 = flags & 2 != 0,
            .self_modify = flags & 1 != 0,
        };
        self.push_word(@bitCast(inst));
    }

    pub fn write_hello(self: *Assembler) void {
        // let i = hello.start
        self.push_instruction(Operation.copy, 0b000);
        self.push_word(0x7f);
        const hello_start = self.addr;
        self.push_word(0);
        const loop_start = self.addr;
        // loop:
        // let char = *i ;;
        self.push_instruction(Operation.deref, 0b110);
        self.push_word(0x80);
        self.push_word(0x7f);
        // jmpz char done
        self.push_instruction(Operation.jumpifz, 0b010);
        const addr_done = self.addr;
        self.push_word(0);
        self.push_word(0x80);
        // i = i + 1
        self.push_instruction(Operation.add, 0b100);
        self.push_word(0x7f);
        self.push_word(0x7f);
        self.push_word(0x1);
        // poke TERM_WRITE char
        self.push_instruction(Operation.poke, 0b010);
        self.push_word(0);
        self.push_word(0x80);
        // jmpz loop 0
        self.push_instruction(Operation.jumpifz, 0b000);
        self.push_word(loop_start);
        self.push_word(0);
        // halt
        self.binary[addr_done] = self.addr;
        self.push_instruction(Operation.setflag, 0b000);
        self.push_word(1);
        const hello = "Hello, World!\n";
        self.binary[hello_start] = self.addr;
        for (hello) |c| self.push_word(c);
        self.push_word(0);
    }
};

File: emulator.c
#include <stdio.h>
#include <stdint.h>
#include <stdbool.h>
#include <stdlib.h>

typedef uint16_t word;
typedef uint8_t byte;

/*
Instruction layout:
__LL _DSS IIII IIII
  L = number of args 0-3
  D = Whether to write back result (to extra provided param)
  S = Whether to read input(s) from an address (1) or immediate (0)
  I = Which of 256 instructions to perform in between.
*/
enum {
    ZERO_PAGE_SIZE = 1 << 8, // First 256 bytes are "registers" on stack
    MEMSIZE = (1 << 16) - ZERO_PAGE_SIZE, // Rest is in heap
    IBIT_READADDR1 = 1 << 8,
    IBIT_READADDR0 = 1 << 9,
    IBIT_WRITEADDR = 1 << 10,
    IOFFSET_LEN = 12,
};

typedef union {
    struct {
        byte operation: 8;
        bool read_addr1: 1;
        bool read_addr0: 1;
        byte write_addr: 1;
        byte _pad0: 1;
        byte argc: 2;
        byte _pad1: 2;
    };
    word iword;
} Instruction;

enum {
    OP_NONE,
    OP_COPY,
    OP_DEREF,
    OP_ADD,
    OP_JUMP_IF_ZERO,
    OP_POKE,
    OP_HALT,
};

typedef struct {
    word pc;
    word reg_in[3];
    word reg_out;
    word zeropage[255];
    word *memory;
    bool halted;
} CPU;

CPU init_cpu() {
    return (CPU){
      .memory = calloc(sizeof(word), MEMSIZE),
      .halted = false,
    };
}

void free_cpu(CPU cpu) {
    free(cpu.memory);
}

void step_cpu(CPU *cpu) {
    bool debug = false;
    Instruction ix = { .iword = cpu->memory[cpu->pc++] };
    word write_addr;
    if(debug) printf("instruction %08b %02x; ", ix.iword >> 8, ix.operation);
    if (ix.write_addr) {
        write_addr = cpu->memory[cpu->pc++];
    }
    for (int i=0; i<ix.argc; ++i) {
        cpu->reg_in[i] = cpu->memory[cpu->pc++];
        if (debug) printf("param %04x; ", cpu->reg_in[i]);
    }
    if (debug) puts("");
    for (int i=0; i<2; ++i) {
        bool memflag = (i == 0) ? ix.read_addr0 : ix.read_addr1;
        if (memflag) {
            word addr = cpu->reg_in[i];
            cpu->reg_in[i] = cpu->memory[addr];
            if (debug) printf("read %04x from %04x\n", cpu->reg_in[i], addr);
        }
    }
    switch (ix.operation) {
        case OP_NONE: break;
        case OP_COPY: cpu->reg_out = cpu->reg_in[0]; break;
        case OP_ADD: cpu->reg_out = cpu->reg_in[0] + cpu->reg_in[1]; break;
        case OP_JUMP_IF_ZERO: if (cpu->reg_in[1] == 0) cpu->pc = cpu->reg_in[0]; break;
        case OP_POKE: putchar(cpu->reg_in[1]); break;
        case OP_HALT: cpu->halted = true;
    }
    if (ix.write_addr) {
        cpu->memory[write_addr] = cpu->reg_out;
        if (debug) printf("wrote %04x to %04x\n", cpu->reg_out, write_addr);
    }
}

int main() {
    CPU cpu = init_cpu();

    // Ghetto assembler in C lmaooo
    char hello_world[] = "Hello, world!\n";
    char *c = hello_world;
    int i=0;
    while (*c) {
        cpu.memory[i] = hello_world[i];
        ++i, ++c;
    }
    int code_start = i;
    // let i = hello.start ;; hello.start is 0
    cpu.memory[i++] = (1 << IOFFSET_LEN) | IBIT_WRITEADDR | OP_COPY;
    cpu.memory[i++] = 127;
    cpu.memory[i++] = 0;
    // loop:
    int loop_start = i;
    // let char = *i ;;
    cpu.memory[i++] = (1 << IOFFSET_LEN) | IBIT_READADDR0 | OP_NONE;
    cpu.memory[i++] = 127;
    cpu.memory[i++] = IBIT_WRITEADDR | IBIT_READADDR0 | OP_COPY;
    cpu.memory[i++] = 128;
    // i += 1
    cpu.memory[i++] = (2 << IOFFSET_LEN) | IBIT_WRITEADDR | IBIT_READADDR1 | OP_ADD;
    cpu.memory[i++] = 127;
    cpu.memory[i++] = 1;
    cpu.memory[i++] = 127;
    // poke TERM_WRITE char
    cpu.memory[i++] = (2 << IOFFSET_LEN) | IBIT_READADDR1 | OP_POKE;
    cpu.memory[i++] = 0;
    cpu.memory[i++] = 128;
    // jmpz done char
    cpu.memory[i++] = (1 << IOFFSET_LEN) | OP_JUMP_IF_ZERO;
    int write_back_to_done_label = i++;
    // jmpz loop 0
    cpu.memory[i++] = (2 << IOFFSET_LEN) | OP_JUMP_IF_ZERO;
    cpu.memory[i++] = loop_start;
    cpu.memory[i++] = 0;
    // done:
    cpu.memory[write_back_to_done_label] = i;
    cpu.memory[i++] = OP_HALT;

    cpu.pc = code_start;

    /* for (int i=0; i<50; ++i) { */
    /*     printf("0x%04x, ", ((cpu.memory[i] & 0xff) << 8) | cpu.memory[i] >> 8); */
    /* } */
    /* puts(""); */
    /* puts(""); */

    for (int i=0; i<100 && !cpu.halted; ++i) {
        step_cpu(&cpu);
    }
    free_cpu(cpu);
    return 0;
}
